File: functions/shared/phone.py
"""
Phone normalisation — MUST produce byte-identical output to:
  - CYBER/functions/src/services/mychama.service.ts (normalizePhone / localPhone / detectProvider)
  - src/lib/phone.ts (this repo, frontend)

members.phoneNormalized is the collection-group lookup key the WhatsApp bot
depends on. A divergent implementation here silently breaks WhatsApp access
for any member added or edited through the app.
"""

from __future__ import annotations

import re

_AIRTEL_PREFIXES = {"073", "075", "078"}


def normalize_phone(phone: str) -> str:
    """Normalise any Kenyan phone spelling to E.164 (+2547XXXXXXXX / +2541XXXXXXXX)."""
    digits = re.sub(r"\D", "", phone or "")

    if re.fullmatch(r"254[17]\d{8}", digits):
        return "+" + digits
    if re.fullmatch(r"0[17]\d{8}", digits):
        return "+254" + digits[1:]
    if re.fullmatch(r"[17]\d{8}", digits):
        return "+254" + digits

    return ("+" + digits) if digits else ""


def local_phone(e164: str) -> str:
    """Local display form (07XXXXXXXX) for anything shown to a member."""
    digits = re.sub(r"\D", "", e164 or "")
    return "0" + digits[3:] if digits.startswith("254") else digits


def detect_provider(phone: str) -> str:
    """Returns 'mpesa' or 'airtel' based on the local prefix."""
    local = local_phone(normalize_phone(phone))
    prefix3 = local[:3]
    return "airtel" if prefix3 in _AIRTEL_PREFIXES else "mpesa"


def is_valid_kenyan_phone(phone: str) -> bool:
    """Mirrors src/lib/phone.ts::isValidKenyanPhone exactly."""
    import re
    n = normalize_phone(phone)
    return bool(n) and bool(re.fullmatch(r"\+254[17]\d{8}", n))


def normalize_sms_phone(phone: str) -> str:
    """
    254XXXXXXXXX form (no +) — what HostPinnacle expects. Mirrors the PAY
    repo's normalizeSmsPhone() exactly.
    """
    e164 = normalize_phone(phone)
    return e164.lstrip("+")

File: src/lib/phone.ts
/**
 * Phone normalisation — MUST produce byte-identical output to
 * functions/shared/phone.py (this repo) and CYBER's mychama.service.ts.
 * members.phoneNormalized is the collection-group lookup key the WhatsApp
 * bot depends on; a divergent implementation here silently breaks WhatsApp
 * access for any member added or edited through the app.
 */

const AIRTEL_PREFIXES = ['073', '075', '078'];

export function normalizePhone(phone: string): string {
  const digits = String(phone || '').replace(/\D/g, '');

  if (/^254[17]\d{8}$/.test(digits)) return '+' + digits;
  if (/^0[17]\d{8}$/.test(digits)) return '+254' + digits.substring(1);
  if (/^[17]\d{8}$/.test(digits)) return '+254' + digits;

  return digits ? '+' + digits : '';
}

export function localPhone(e164: string): string {
  const digits = String(e164 || '').replace(/\D/g, '');
  return digits.startsWith('254') ? '0' + digits.substring(3) : digits;
}

export function detectProvider(phone: string): 'mpesa' | 'airtel' {
  const local = localPhone(normalizePhone(phone));
  const prefix3 = local.slice(0, 3);
  return AIRTEL_PREFIXES.includes(prefix3) ? 'airtel' : 'mpesa';
}

export function isValidKenyanPhone(phone: string): boolean {
  return normalizePhone(phone) !== '' && /^\+254[17]\d{8}$/.test(normalizePhone(phone));
}
